### PR TITLE
Multi-color text with color encodings

### LINKED ISSUE #2019: Allow text coloring and add edge to text
## 🚀 Feature
- [ ] Ability to pass color array to text (to color each text label differently)
- [ ] Ability to set an edge (color and thickness) to text

## Motivation

We'd like to plot text color according to some values. Also, and edge around the text would be useful

## Pitch

We are working toward interfacing anndata https://github.com/theislab/anndata with napari for sptial transcriptomics data, and would like to get text visualized like this (color correspond to cluster coloring):
image

## Alternatives
For now, it only uses the first color of the array, if an array is passed to text:
```python
_text_color
# array([[0.        , 0.        , 0.        ],
#       [0.        , 0.        , 0.        ],
#       [0.        , 0.        , 0.        ],
#       ...,
#       [0.        , 0.        , 0.        ],
#       [0.15294118, 0.61960784, 0.40784314],
#       [0.        , 0.        , 0.        ]])
text = {
	"text": "{clusters}",
	"size": 24,
	"color": _text_color,  
	"anchor": "center",
	"blending": "translucent",
}
```
image

## Additional context

There is a quick fix wrt coloring (see PR) but would be happy to have inputs on whether it would be possible to also have an edge for the text

pinging @michalk8 with whom we are working on this.

Thank you!

### LINKED ISSUE #4365: Use ColorEncoding in TextManager
The goal here is to use the ColorEncoding described in the [style encoding design doc](https://hackmd.io/Ybil_epnQPiBczPcbGRqCw?view) and implemented in #4357 in TextManager.

This enables some of the functionality described in #2019 and also acts as a good testing ground for color encodings in napari before we use them for things like `Points.face_color`.

In an architecture working group meeting, we decided that it would be OK to use `TextManager.color` for the new `ColorEncoding` field. This will break anyone currently reading `Layer.text.color` and expecting an `np.ndarray[(4,), np.float32]`, but our guess is there aren't too many of those people and the fix is fairly trivial (i.e. read `Layer.text.color.constant` instead).

If we want to avoid breakage, an alternative is to deprecate `color` and add `face_color` instead to store `ColorEncoding`, given that #2019 also wants to add a distinct edge color.

### PR DESCRIPTION
**Pull request by andy-sweet**
# Description

This PR enables multi-color text in the points and shapes layers by using a `ColorEncoding` in `TextManager`.

For a wider context of the style values from properties/features, see the design doc: https://hackmd.io/Ybil_epnQPiBczPcbGRqCw?view

#### Makes the `color_encoding` module public

This is the first public usage of the `color_encoding` module, so we need to make it public here.

#### Change `TextManager.color` to be a `ColorEncoding`

This is a breaking change that affects reading `TextManager.color`. Previously you could only pass a single color

```python
In [1]: points.text.color = 'red'

In [2]: points.text.color
Out[2]: array([1., 0., 0., 1.], dtype=float32)
```

which would be coerced to an RGBA array. Now that `color` is a `ColorEncoding`, we have

```python
In [3]: points.text.color = 'red'

In [4]: points.text.color
Out[4]: ConstantColorEncoding(constant=array([1., 0., 0., 1.], dtype=float32))
```

where you need to read `points.text.color.constant` to get the same array.

#### Adds an example of multi-color text

This was copied from `add_points_with_text.py`, but encodes text color from features in the same way as point face colors does. This should become more concise when points also uses color encodings.

## Type of change
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
Closes #4365
Probably closes #2019, though we should open a new issue for allow a text edge/outline color (even though vispy doesn't currently support it).

# How has this been tested?
- [x] newly added tests pass with my change
- [x] all existing tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by jni** [comment]
OMG! Slow down @andy-sweet I can't keep up! 😂 👏 👏

**Comment by andy-sweet** [comment]
Well, I had this mostly prepped in advance. @jni : I'll let you take a breather for now and have assigned @kne42 to take a first pass at this.

File: napari/layers/utils/text_manager.py
Comment (kne42) [inline napari/layers/utils/text_manager.py]: would be nice to send a message detailing the equivalent setter for `face_color` e.g. `manager.color = cyan` -> `manager.face_color = {'constant': cyan}`

**Review by kne42 — COMMENTED** [review]
Overall, excellent code quality and nice handling of the deprecations.

A few big picture questions (apologies if I'm missing context on these):
1. what's the reasoning behind renaming `color` to `face_color`?
2. is there a specific reason why constant colors must use the formatting of `{'constant': 'red'}` instead of having the setter coerce it into a valid `ColorEncoding`?

**Comment by andy-sweet** [comment]
Thanks for the review!

Both high level questions are important, so thanks for raising them.

> 1. what's the reasoning behind renaming `color` to `face_color`?

I think the biggest reason is that it makes the deprecation of `color` easier, without breaking anyone who was reading `TextManager.color` and expecting to get an RGBA array. Another less important reason is that in #2019 there is at least a desire to have a separate `edge_color` for text - we can't support that in vispy yet, but may be able to in the future.

That being said, I think @jni was OK with breaking that usage of `color` as the usage is likely limited and the fix is straightforward (i.e. `TextManager.color.constant` instead). It might just be better to go with that and avoid the extra work needed for the deprecation here.

> 2\. is there a specific reason why constant colors must use the formatting of `{'constant': 'red'}` instead of having the setter coerce it into a valid `ColorEncoding`?

This is a bigger/wider question around how we coerce values passed to style encodings. The general logic across all encoding types is that if you pass a `str`, then we assume the style value is mapped directly from a feature (e.g. `DirectColorEncoding`, `DirectStringEncoding`). I.e. the logic of coercion is dependent only on the type of the value passed, which is nice and simple.

We could consider making the coercion also dependent on the value passed in when it's a string. In this case, if we can successfully parse a single color from a `str` value passed, then we could return a `ConstantColorEncoding` and retain the behavior from before. E.g.

```python
text_manager.color = 'red'
text_manager.color # => ConstantColorEncoding(constant=array([1, 0, 0, 1]))
```

As I write this, I realize we already do something similar for string encodings. That is when the value passed is a valid format string, we coerce it into a `FormatStringEncoding`. And I've also considered parsing strings generally in an altair-like way, so that `text_manager.color = 'class:n'` would be coerced into a `NominalColorEncoding(feature='class')`.

So I think it would be pretty valid to make the color encoding coercion return a constant color encoding when the string value represents a single color. Especially as that could be a relatively common usage. What do you think? 

Note that one thing we initially wanted to avoid is making the coercion also dependent on `features`. For example, if we pass a string, then we might check is that string is a column name in the feature table and return a `DirectColorEncoding` if that's the case and try to return `ConstantColorEncoding` otherwise. That's closer to altair, but makes the coercion logic more complicated and also means we need to store `features` as state in `TextManager` and other things that own encodings.

File: napari/layers/utils/text_manager.py
Comment (andy-sweet) [reply napari/layers/utils/text_manager.py]: Great idea! I'm going to hold off on making the change until we're more certain on introducing `face_color` and the new coercion behavior.

**Comment by jni** [comment]
> So I think it would be pretty valid to make the color encoding coercion return a constant color encoding when the string value represents a single color. Especially as that could be a relatively common usage. What do you think?
> 
> Note that one thing we initially wanted to avoid is making the coercion also dependent on features. For example, if we pass a string, then we might check is that string is a column name in the feature table and return a DirectColorEncoding if that's the case and try to return ConstantColorEncoding otherwise. That's closer to altair, but makes the coercion logic more complicated and also means we need to store features as state in TextManager and other things that own encodings.

Yeah, that was the whole discussion, that if there is a column called 'red', then *without* `{'constant': 'red'}`, it is ambiguous and value-dependent what `color='red'` means. Personally I really like the idea of (a) disambiguating this, and (b) making the feature mapping the default value. (ie making color=column_name the happy path). But yeah, @andy-sweet, maybe this is one of those decisions that would really benefit from a NAP, so that both the decision rationale and discussion can be captured, and so that more people can partake in the decision.

**Comment by kne42** [comment]
> I think the biggest reason is that it makes the deprecation of `color` easier, without breaking anyone who was reading `TextManager.color` and expecting to get an RGBA array.

tbh I don't think that's a good enough reason to change the name to something that isn't as intuitive, especially if not many people are going to be affected by a breakage. also re: the points you bring up later this may even not break depending on how we handle things

> Another less important reason is that in #2019 there is at least a desire to have a separate `edge_color` for text - we can't support that in vispy yet, but may be able to in the future.

i think `face_color` and `edge_color` are a bit confusing in the context of text, id think `color` and `outline_color` would be more intuitive. however, im not the target userbase so i will defer to people who understand them better

> This is a bigger/wider question around how we coerce values passed to style encodings. The general logic across all encoding types is that if you pass a `str`, then we assume the style value is mapped directly from a feature (e.g. `DirectColorEncoding`, `DirectStringEncoding`). I.e. the logic of coercion is dependent only on the type of the value passed, which is nice and simple.

could you provide additional context on this pls?

> We could consider making the coercion also dependent on the value passed in when it's a string. In this case, if we can successfully parse a single color from a `str` value passed, then we could return a `ConstantColorEncoding` and retain the behavior from before. E.g.
> 
> ```python
> text_manager.color = 'red'
> text_manager.color # => ConstantColorEncoding(constant=array([1, 0, 0, 1]))
> ```
> 
> As I write this, I realize we already do something similar for string encodings. That is when the value passed is a valid format string, we coerce it into a `FormatStringEncoding`. And I've also considered parsing strings generally in an altair-like way, so that `text_manager.color = 'class:n'` would be coerced into a `NominalColorEncoding(feature='class')`.

I like this a lot better. imo the `{ 'cat': 'kitten' }` format is too verbose unless you are passing in multiple things like `{ 'cat': 'kitten', 'dog': 'puppy' }`

> So I think it would be pretty valid to make the color encoding coercion return a constant color encoding when the string value represents a single color. Especially as that could be a relatively common usage. What do you think?

I like this. theres multiple ways to handle this, we could store the initial value passed to `color` and spit it back out, or we could build-in some simplification into `ColorEncoding` like u were suggesting and then add in methods for equivalency and `__repr__` to make it seamless on the user end

> Note that one thing we initially wanted to avoid is making the coercion also dependent on `features`. For example, if we pass a string, then we might check is that string is a column name in the feature table and return a `DirectColorEncoding` if that's the case and try to return `ConstantColorEncoding` otherwise. That's closer to altair, but makes the coercion logic more complicated and also means we need to store `features` as state in `TextManager` and other things that own encodings.

I think this should be more explicit so you're checking for something like `'column:red'` if u want to target a specific feature (tbh i dont understand style features very well/at all)

**Comment by jni** [comment]
Ok so we're at the SF hackathon and we are thinking that the July release will be 0.5 (:rocket:) and we are going to Move Fast and Break Things, so maybe we can revisit the color->face_color decision?

But I do want to push back a bit on the 'column:colname' idea. String-based encoding is hard to get right at the parsing step, and you can add unlimited things, e.g. 'column:colname:Q'. So I still prefer a dictionary that corresponds to specific classes of encodings as a general rule.

**Comment by andy-sweet** [comment]
I updated this PR in two ways based on some in-person feedback.

1. Removed the new `face_color`, and reused the existing the `color` field, breaking its current behavior.
2. Do not coerce a color encoding from a string.

Change (1) is in the spirit of breaking things for version 0.5 and I think is good.

I'm less sure about (2) and I think it would be better to do something with a string. I'm planning to write up the options and their implications for encodings in general, so we can get some consensus there.

**Review by jni — APPROVED** [review]
This looks great @andy-sweet! Regarding your second question (removing string-only setting), as we [discussed in the architecture WG meeting](https://napari.zulipchat.com/#narrow/stream/296574-working-group-architecture/topic/Coercing.20style.20encodings.20from.20strings/near/283822290), I think it's ok to leave this off as a first pass, and we can decide how to implement strings later pending broader community input. It's easier to expand the API than to shrink it.

**Comment by jni** [comment]
This needs an update with main.

**Comment by andy-sweet** [comment]
This is on a temporary hold while we discuss/poll some existing color coercion behavior: https://napari.zulipchat.com/#narrow/stream/212875-general/topic/Vote.3A.20colors.20from.20strings

Please don't merge yet!

**Comment by andy-sweet** [comment]
Based on [the discussion/vote on Zulip](https://napari.zulipchat.com/#narrow/stream/212875-general/topic/Vote.3A.20colors.20from.20strings), I decided to support parsing constant color encodings from single color names like `text.color = 'red'`, which reduces the number of changes in this PR.

@jni and @kne42 please take another look when you get a chance.

I also defined `assert_colors_equal` in `napari._tests.utils` as we use that in more than place now.

File: napari/layers/utils/_tests/test_text_manager.py
Comment (kne42) [inline napari/layers/utils/_tests/test_text_manager.py:594]: is cyan the default colour or something? if so pls explain in a comment

File: napari/layers/utils/_tests/test_text_manager.py
Comment (kne42) [inline napari/layers/utils/_tests/test_text_manager.py]: nit: pls add comments to tests like these to specify expected behaviour e.g. "first three features should be red, green, and blue as specified, while the remaining two default to yellow"

File: napari/layers/utils/_tests/test_text_manager.py
Comment (kne42) [inline napari/layers/utils/_tests/test_text_manager.py:692]: nit (no change needed): what's with the naming this as `paste`? imo `append` would make more sense as it's specifically adding it to the end of the list of features

File: napari/layers/utils/color_encoding.py
Comment (kne42) [inline napari/layers/utils/color_encoding.py:50]: ah i see now that it is indeed the default color. in that case, in the tests above please specifically use this constant where appropriate (assuming i understood the tests correctly)

**Review by kne42 — APPROVED** [review]
excellent work! some small comments but otherwise LGTM

File: napari/layers/utils/_tests/test_text_manager.py
Comment (andy-sweet) [reply napari/layers/utils/_tests/test_text_manager.py:594]: Good catch! It's actually the fallback color that's used when the encoding fails. Instead of adding a comment and/or using the constant (which is the default fallback), I updated the test name to indicate the expected behavior here and specified the fallback in the input so that the expected values are more clear.

File: napari/layers/utils/_tests/test_text_manager.py
Comment (andy-sweet) [reply napari/layers/utils/_tests/test_text_manager.py:692]: It's for consistency with [`Points._paste_data`](https://cs.github.com/napari/napari/blob/2dc5aa659f875c353bfbde3b20d8f07a664ed8a8/napari/layers/points/points.py?q=points#L1851) and [`Shapes._paste_data`](https://cs.github.com/napari/napari/blob/2dc5aa659f875c353bfbde3b20d8f07a664ed8a8/napari/layers/shapes/shapes.py?q=shapes#L2896).

I don't feel too strongly about the name here, but as it's not public yet and this is just a nit, I'm going to leave this as is for now.

File: napari/layers/utils/_tests/test_text_manager.py
Comment (andy-sweet) [reply napari/layers/utils/_tests/test_text_manager.py]: I usually interpret explanatory/implementation comments in tests as a bit of a code smell, so I try to avoid them unless they're really needed to explain complex or confusing behavior (which is sometimes hard to avoid). But I do like including expected behavior in test names themselves, because that's something you see in test outputs, so it makes it clearer what behavior broke when this test fails. So I did that here.

Otherwise, I think the test is pretty readable as it is and the comment above doesn't add a huge amount in terms of explanatory power after updating the test name. Obviously, I have a pretty biased perspective as the author of the test and implementation, so definitely happy to add a comment if really desired, but as this is a nit I'll leave it for now.

**Review by andy-sweet — COMMENTED** [review]
Thanks for the review @kne42 ! I didn't address all your comments exactly, but tried to make changes that captured their spirit and tried to explain my reasoning. Feel free to push back if you still want to pick those specific nits.

**Review by jni — APPROVED** [review]

**Pull request merged**
